### ISSUE TITLE
[proposal] Middleware API

### ISSUE THREAD

**Issue by dmitrykiselev27** [aws/chalice#1258]
## Overview
This is a proposal for adding middleware support to the chalice framework. It outlines the middleware API and its usage for building REST APIs.
Middlewares allow to modify:
1. The request before it will be processed by the view by doing the modifications of `app.current_request` instance.
2. The response which will be returned to the user.

## Middleware API
The middleware interface is inspired by the middleware implementation in Django and Flask.
In general, middleware is a class derived from the base `Middleware` class which provides the interface which must be implemented by the concrete middleware.
```
from chalice import Middleware

class ExampleMiddleware(Middleware):
        def __call__(self, *args, **kwargs):
            # do something with app.current_request here
            response = super().__call__(*args, **kwargs)
            # modify response here
            return response
```
It defines the `__call__` method which will will call the view function or a next middleware in the chain. 

## Middleware registration
As middleware is global state objects they must be stored as part of the Chalice application instance.
This proposal introduces the `app.middlewares` attribute - iterable (list by default) with all middlewares enabled for the application.
```
    from chalice import Chalice, Middleware

    app = Chalice(app_name='middleware-demo')

    class ExampleMiddleware(Middleware):
        def __call__(self, *args, **kwargs):
            # do something with app.current_request here
            response = super().__call__(*args, **kwargs)
            # modify response here
            return response


    app.middlewares = [ExampleMiddleware]
```

**Comment by muriloviana** [comment]
Hey everyone, this implementation will be very welcome. Do we have any update?

**Comment by gmachado-janrain** [comment]
+1

**Comment by dmitrykiselev27** [comment]
@stealthycoin is there a chance to get feedback for this one?

**Comment by jamesls** [comment]
Hi, I'm currently looking into this and am trying to consolidate several features.  After looking over the various proposals in this space, I like the idea of adding middleware, but I have a few suggestions:

1. I'd prefer to stay consistent with how Chalice's decorator based API so I'd like some sort of `@app.middleware` decorator.  This is also closer aligned to Flask's `@app.before/after_request` decorators and I think will work better with Blueprints. Along these lines, while people can always write classes if they prefer, I think we should make it simple for people to write function-based middleware.
2. I'd prefer a solution that works for all event types (e.g. all our decorators).  So while I believe we should have the ability to add middleware to REST APIs, we should also support `@app.on_s3_event, @app.on_ws_message`, etc.  I think Chalice is unique in the sense that it supports more than just REST APIs so we have the opportunity to create an API for Lambda-based middleware.
3. (expanding on 2) There should be some way to filter out different events so that if you wanted middleware for only a specific event type (e.g. SQS) you'd be able to specify that somehow.

I'm thinking something similar to Django's middleware, but more along the API of how fastapi/starlette do it with decorators.  Thoughts?

**Comment by dmitrykiselev27** [comment]
@jamesls thanks, I'll look forward to your proposal

**Comment by jamesls** [comment]
Resolved via #1509
